### DISCUSSION TITLE
Umgang mit HB (Homebrew) Geräten

### DISCUSSION THREAD

**Discussion by Baxxy13** [SukramJ/aiohomematic#276] [Q&A]
Wer nicht nur Homematic-Freund sondern auch Bastler ist kennt sicher die HB-xxx Aktoren / Sensoren aus der Selbstbau-Ecke.

Ich war erstaunt und erfreut das z.B. der `HB-UNI-Sensor1` fast korrekt dargestellt wird. 
Fast, weil z.B. Datenpunkte / Parameter, die so normalerweise auf einer CCUx / RM nicht vorkommen natürlich auch nicht korrekt behandelt werden.

Im Fall meines `HB-UNI-Sensor1` sind das die Datenpunkte / Parameter: `"Taupunkt"` und `"Abs_Luftfeuchte"`
`"Taupunkt"` ist die Taupunkttemperatur in °C
`"Abs_Luftfeuchte"` ist die Absolute Luftfeuchte in g/m³

Aktuell habe ich die Darstellung mittels "customize" ausgebessert...
```
  customize:
    # UNISENSOR-3
    sensor.klima_sensor_bx_unisens003_1_klimadaten_abs_luftfeuchte:
      device_class: humidity
      unit_of_measurement: g/m³
    sensor.klima_sensor_bx_unisens003_1_klimadaten_taupunkt:
      device_class: temperature
      unit_of_measurement: °C
```
... und das sieht dann so aus:
![HA_RM_Unisensor_Customize](https://user-images.githubusercontent.com/54898249/152835042-f1ac65dc-27fb-4339-a883-8d392dd472be.JPG)
(rot markiert die customisierten Parameter)

-----------------------------------------------

Die primäre Frage wär nun, ob es gewünscht ist solche speziellen Datenpunkte / Parameter mit in die Integration aufzunehmen, so das diese nach der Einrichtung gleich korrekt und vollständige angezeigt werden.

Hierfür könnte ich dann im Homematic-Forum bei den Entwicklern der Geräte um Auflistung der speziellen Datenpunkte / Parameter bitten und diese zusammentragen.

Bei der absoluten Luftfeuchte bin ich aber auch schon auf ein Problem gestoßen. HA kennt wohl weder `g/m³` noch `g/kg`. 
Siehe: [Concentration units](https://github.com/home-assistant/core/blob/96c4e33b2427308307f2983c446617512075c39f/homeassistant/const.py#L587-L593)
Ich kann aber nicht beurteilen ob das problematisch ist.

-----------------------------------------------

Auch hier habe ich wieder ein "Spezial-Problem", oder Wunsch... 
(HM-xxx vs. HB-xxx)

Weil ich die Tasten eines **HM**-LC-Bl1PBU-FM für andere Zwecke nutzen wollte (sowie zum testen und spielen) habe ich den zu einem **HB**-LC-Bl1PBU-FM umgeflasht. Technisch gesehen ist das Gerät vollkommen identisch mit der Ausnahme, das sich die Tasten für DV's mit anderen Aktoren nutzen lassen. Also Kanal:1 ist der Rollladenaktor, :2 & :3 die Tasten. Die Tasten sind hier aber irrelevant da sie nur für DV's funktionieren und nicht von der Zentrale aus.
Vermutlich durch die andere Bezeichnung wird der Aktor nun nicht als Rollladenaktor erkannt. 
![HA_RM_HB-BL1_custom](https://user-images.githubusercontent.com/54898249/152847772-f783c285-2c20-4be2-b5c2-79982bd6bc6d.JPG)

Lässt sich da was machen? 
Dann könnte ich auch den offenen Test mit `"DIRECTION"` durchführen.

**Comment by SukramJ** [comment]
Wenn du bei mehreren Punkten Nummern zu schreibst lässt sich besser antworten.

Aktuell reagieren wir bei den custom_entities auf den device_type. Ist der device_type also HM-LC-Bl1PBU-FM, dann wird ein Cover erstellt.  Beim HM-LC-Bl1PBU-FM erwarten wir dann auf Kanal 1 einen Datenpunkt LEVEL.
Die Tasten werden separat behandelt und spielen im Bereich Cover keine Rolle. 
Du kannst aber jede Taste über einen Device Trigger in einer Automation verwenden.
Grundsatzfrage: Ist ein HM-LC-Bl1PBU-FM bei den Datenpunkten komplett identisch zu einem HB-LC-Bl1PBU-FM? 
Worin liegt der Unterschied?
Heissen nach dem umfassen immer HB-XXXX?

Zu den Sensoren:
Es macht keinen Sinn einen "persönlichen" Baxxy13-Taupunkt Datenpunkt in custom_homematic aufzunehmen.
Wenn aber es aber viele Geräte gibt, die diesen Datenpunkt haben, dann kann man schon darüber nachdenken.

Wenn es ein einfaches Mapping zwischen Parameter und Unit gibt ist das kein Problem.

ABS_LUFTFEUCHTE -> g/m³
TAUPUNKT -> °C bzw. TEMP_CELSIUS

Bereits in HA gepflegte Konstanten wie TEMP_CELSIUS sind dabei natürlich zu verwenden, es können aber auch eigene Einheiten angegeben werden.

**Comment by Baxxy13** [comment]
A1: 
zu den Sensoren:
Du hattest mir mal einen Link präsentiert [sensors.py](https://github.com/danielperna84/pyhomematic/blob/899a28c6036ea799cfb5513a65e37135556c60a7/pyhomematic/devicetypes/sensors.py#L1325-L1342)
Da sind ja schon einige der "bekanntesten" HB-xxx Geräte aufgeführt. So auch der erwähnte `HB-UNI-Sensor1`. Da fehlen halt nur die Definitionen für `"Taupunkt"` und `"Abs_Luftfeuchte"`

Es geht mir auch nicht um persönliche Datenpunkte. Das würde jedem zugute kommen der z.B. den `HB-UNI-Sensor1` einsetzt und die Integration nutzt. Wie viele User das wären kann ich nicht sagen. Aber ob z.B. `"Taupunkt"` noch von anderen HB-xxx Sensoren genutzt wird lässt sich rausbekommen.

A2: 
spezielle Einheiten:
Ok, da gibt es also keine Probleme, sehr gut.

B1: 
HM-LC-Bl1PBU-FM vs. HB-LC-Bl1PBU-FM:
Die Kanal / Datenpunktstruktur ist bis zu Kanal:1 vollkommen identisch. 
Der HB-xx hat dann noch Kanal:2 und Kanal:3 als Tasterkanal. Diese gibt es beim Original nicht, da die Tasten bei HM noch fest mit dem Aktor "verheiratet" sind. So gesehen ist es fast ein IP Aktor wo es auch separate Tasterkanäle gibt.

B2:
Benamung der HB Geräte:
Soweit ich weiß heißen alle Geräte die es so nicht im Original als HM Aktoren / Sensoren gibt HB-XX.
Ehemalige original Geräte die mit alternativer Firmware erweitert wurden dann auch HB-XX.
Beispiele die ich hier habe:
HB-LC-Sw1PBU-FM
HB-LC-Sw2PBU-FM
HB-LC-Bl1PBU-FM

**Reply by SukramJ** [reply]
A1 sensor.py ist aus dem alten Projekt. In HAHomematic werden keine Sensoren definiert, da definieren  wir nur wie ein Datenpunkt/Parameter"[hübsch](https://github.com/danielperna84/custom_homematic/blob/devel/custom_components/hahm/entity_helpers.py)" angezeigt wird.
Ist "Taupunkt" und "Abs_Luftfeuchte" die richtige Schreibweise (groß/klein) ?

B2:
Lösungsvorschlag für custom_entities: wir definieren nur Original-Geräte (HM-xxx), suchen aber auch nach HB-xxx.
Solange sich die Kanäle von HM-xxx bei einem HB-xxx nicht verschieben gibt es keine Probleme. angehängte Taster sind erstmal egal.

**Reply by SukramJ** [reply]
Um die Sache rund zu machen solltest Du die Gerätebeschreibungen für pydevccu erzeugen. 
Daniel hat das [hier](https://github.com/danielperna84/hahomematic/discussions/127#discussioncomment-1904703) beschrieben.
Es gibt in HA eine  Service `Homematic(IP) Local: Export device definition`, der die benötigten Dateien in den Order {ha-config}/hahm/... exportiert.
Am Besten gleich bei [pydevccu](https://github.com/danielperna84/pydevccu) als PR einstellen. Dann kann ich die Geräte, zumindest statisch, zum Testen verwenden.

**Reply by SukramJ** [reply]
0.31.1 sollte HB-LC-Bl1PBU-FM als Cover anzeigen

**Reply by Baxxy13** [reply]
A1: 
Ja, die Schreibweise passt so. Hier nochmal ein Screenshot. 
![HA_UNISensor1_special_Params](https://user-images.githubusercontent.com/54898249/152875809-5ecc40cf-c290-4a3c-b5a5-45e69562e233.JPG)

B2: 
Kling gut. Hast du das in der 0.31.1 schon so umgesetzt? Teste ich gleich mal.

C:
Erzeugen der Gerätebeschreibungen.
Ich schaue mal ob ich das hinbekomme, habe mich bisher noch nicht mit den HA-Services befasst.

**Reply by Baxxy13** [reply]
B2:
Der HB-LC-Bl1PBU-FM wird jetzt als Cover integriert und ist bedienbar.
Dafür sind jetzt die unignore-Parameter sowie LEVEL deaktiviert.
![HA_RM_HB-BL1_custom_neu](https://user-images.githubusercontent.com/54898249/152883533-cb629702-f49d-4a34-9e00-7be86b7ea7fa.JPG)

C: 
Die Erzeugung der Gerätebeschreibungen war einfach. Gibt es noch eine andere Option dir die zukommen zu lassen?
Mit PR's kenne ich mich gar nicht aus.

**Reply by SukramJ** [reply]
B2 Mit 0.31.2 wird  die unignore Datei wieder richtig eingelesen. Da hatte ich zu viel umbenannt.
LEVEL kann gelöscht werden, da LEVEL in Cover enthalten ist.

**Reply by Baxxy13** [reply]
B2:
sieht jetzt gut aus.
![HA_RM_HB-BL1_custom_neu_ok](https://user-images.githubusercontent.com/54898249/152957712-b463a27c-d621-473d-80e3-d77bd7e7255a.JPG)

A und C:
Die Definitionen für `HB-UNI-Sensor1` und `HB-LC-Bl1PBU-FM` sind als PR raus.

`HB-LC-Sw1PBU-FM` und `HB-LC-Sw2PBU-FM` hole ich bei Gelegenheit mal aus der Bastelkiste, lerne die an und reiche die Definitionen nach.

**Reply by SukramJ** [reply]
"Taupunkt" und "Abs_Luftfeuchte" sollten mit 0.31.3 richtig angezeigt werde.
Ich verwende hierbei aber nicht die Namen, sondern die mitgelieferten `Units`.

**Reply by Baxxy13** [reply]
Klappt perfekt und sieht jetzt ohne "customize" genauso aus wie auf meinem Screenshot im 1.Post.

Sehr gut und besten Dank!
